### ISSUE TITLE
Can we move INPUT_direct and FORWARD_direct to the top of INPUT/FORWARD

### ISSUE THREAD

**Issue by divereigh** [firewalld/firewalld#44]
I am running [rtpengine](https://github.com/sipwise/rtpengine) which runs a kernel module that essentially passes packets back out without it having to go through the user-space stack (for performance). However it needs to get in before the conntrack rule (otherwise the conntrack rule just accepts the packet and passes it into userspace, bypassing the kernel module).

I have tried using the --direct interface to add in a rule above the conntrack (in the INPUT chain), however it just gets added to the INPUT_direct chain instead, which comes after conntrack. I have had to resort to creating a new chain (INPUT_prefilter) using the --direct --passthrough. However this seems like a hack. Also anything done with --passthrough has to be --permanent otherwise it won't survive a --reload (is that a bug?).

So I see two solutions:
1) Move INPUT_direct so that it is the first in the INPUT chain
2) Create a INPUT_prefilter chain (as I have done) that is the first in the INPUT chain.

Then for completeness we should do the same with FORWARD.

I am happy to produce a patch to do either.

Thoughts?

**Comment by t-woerner** [comment]
I thought about addition something similar as the prefiter flag also some time ago. But his would require to change the D-Bus interface of the direct interface in firewalld.

There is a solution without the need to change the interface: Use a passthrough rule to add a rule in the first line of the INPUT, .. chain. In this rule you can also jump into chains you have created with the --add-chain option in the same ipv and table.

**Comment by divereigh** [comment]
I have to admit I don't enough about the D-Bus interface, but I'm guessing that rules out option 1. However is there any reason we can't do 2 - i.e. create a standard chain that runs first up before any conntrack.

I understand that you can use --passthrough - that is essentially what I say in the second paragraph above. The issue is that this would be hard to do in any sort of automatic install - for example in an init script. The --passthrough option only works if use --permanent (if you don't, your changes will be lost when you do a --reload), it is also hard to do a remove (essentially you have to do it by rule number - hoping that it is still number 1 in the INPUT chain).

It would be better to do it as a standard chain with a known (pre-defined) name. What happens if there are multiple kernel modules that wan't to get ahead of conntrack?

**Comment by mstrhh** [comment]
There is another issue with the conntrack entry before any other chain:

I cannot block intruders while they are holding a connection! For example a drop entry in IN_public_deny will only take effect once the hacker starts a new connection.

 This looks like a major security problem to me.

**Comment by marcosfrm** [comment]
@t-woerner any news on this?

**Comment by marcosfrm** [comment]
Ping?

**Comment by erig0** [comment]
We're currently working on an nftables backend, which would be a good time to make a change like this. That being said, I don't think it's likely to happen with the iptables backend for the reasons @t-woerner mentions above. The passthrough solution is a bit hacky, but it's the best option right now.

**Comment by josephtingiris** [comment]
FWIW, here's one way to (manually) fulfill the description of this ticket.  It effectively "moves" INPUT_direct & FORWARD_direct to the top of their respective chains.

```sh
# firewall-cmd --direct --passthrough ipv4 -D INPUT -j INPUT_direct                                                          
# firewall-cmd --direct --passthrough ipv6 -D INPUT -j INPUT_direct                                                          
# firewall-cmd --direct --passthrough ipv4 -D FORWARD -j FORWARD_direct                                                          # firewall-cmd --direct --passthrough ipv6 -D FORWARD -j FORWARD_direct                                                          
# firewall-cmd --direct --passthrough ipv4 -I INPUT 1 -j INPUT_direct
# firewall-cmd --direct --passthrough ipv6 -I INPUT 1 -j INPUT_direct
# firewall-cmd --direct --passthrough ipv4 -I FORWARD 1 -j FORWARD_direct
# firewall-cmd --direct --passthrough ipv6 -I FORWARD 1 -j FORWARD_direct
```
Making the above --permanent produces the following `direct.xml`, which works, with the caveat that firewalld does _not_ actually execute the -D arguments.  As it is, it's an imperfect workaround.

```xml
<?xml version="1.0" encoding="utf-8"?>
<direct>
  <passthrough ipv="ipv4">-D INPUT -j INPUT_direct</passthrough>
  <passthrough ipv="ipv6">-D INPUT -j INPUT_direct</passthrough>
  <passthrough ipv="ipv4">-D FORWARD -j FORWARD_direct</passthrough>
  <passthrough ipv="ipv6">-D FORWARD -j FORWARD_direct</passthrough>
  <passthrough ipv="ipv4">-I INPUT 1 -j INPUT_direct</passthrough>
  <passthrough ipv="ipv6">-I INPUT 1 -j INPUT_direct</passthrough>
  <passthrough ipv="ipv4">-I FORWARD 1 -j FORWARD_direct</passthrough>
  <passthrough ipv="ipv6">-I FORWARD 1 -j FORWARD_direct</passthrough>
</direct>
```

Without hacking, the result is 'permanent' INPUT & FORWARD chains with duplicate rules.  The direct.xml INPUT_direct & FORWARD_direct jumps are successfully reinserted at the top of each chain, but the originals still exist further down in the rule set too.  

It would be _very_ helpful if direct _passthrough_ does what it implies.  It's misleading and took a bit of spelunking to figure out that I was being saved from myself.   My expectation was that firewalld would pass through ALL direct arguments.   But, I was wrong.

```
firewalld[3301]: WARNING: INVALID_PASSTHROUGH: arg '-D' is not allowed
```

IMHO, passthrough should at least support "-D", "--delete".  Making that happen takes minimal effort and could effectively resolve many issues for many users.  The fact that direct.xml gets processed last allows for workarounds like this to function as expected & allows for fine tuning.

[I tested this change](https://github.com/firewalld/firewalld/pull/291/files), with the above XML, and it works as expected. 

The same logic works for the other chains too, e.g. OUTPUT

HTH

**Comment by erig0** [comment]
This was addressed with the new nftables backend (v0.6.0). Direct rules have higher precedence and occur before any other firewalld rules. For more info please read the blog post: https://firewalld.org/2018/07/nftables-backend
